### PR TITLE
[READY] Gives Stacking Machines and their consoles machine boards

### PR DESCRIPTION
**Pull request by ghost**
:cl: iksyp
add: stacking machines and their consoles no longer dissapear into the shadow realm when broken
add: you can link the stacking machine and its console by using a multitool
/:cl:
TODO:
- [x] Make the console and machine linkable.
- [x] Break up the engineering nodes, since I'm a liar.
- [x] Change the recipes to be a little less shit.

### CODE REVIEW

File: code/modules/research/techweb/all_nodes.dm [hunk 1]
```diff
@@ -93,7 +93,7 @@
 	description = "A refresher course on modern engineering technology."
 	prereq_ids = list("base")
 	design_ids = list("solarcontrol", "recharger", "powermonitor", "rped", "pacman", "adv_capacitor", "adv_scanning", "emitter", "high_cell", "adv_matter_bin",
-	"atmosalerts", "atmos_control", "recycler", "autolathe", "high_micro_laser", "nano_mani", "mesons", "thermomachine", "rad_collector", "tesla_coil", "grounding_rod", "apc_control", "cell_charger")
+	"atmosalerts", "atmos_control", "recycler", "autolathe", "high_micro_laser", "nano_mani", "mesons", "thermomachine", "rad_collector", "tesla_coil", "grounding_rod", "apc_control", "cell_charger", "stack_console", "stack_machine")
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/research/techweb/all_nodes.dm]: We need to break this up eventually

Comment (iskyp) [reply code/modules/research/techweb/all_nodes.dm]: eh, probably, but i'm not gonna do it.

**Comment by PKPenguin321** [comment]
can you link them if you rebuild them?

**Comment by iskyp** [comment]
i'm not actually sure, i'll test it when i get home

**Comment by Jalleo** [comment]
No. This is map defined stuff I believe.

**Comment by ExcessiveUseOfCobblestone** [comment]
just have the computer and machine key editable with a multitool or something.

File: code/modules/mining/machine_stacking.dm [hunk 2]
```diff
@@ -7,6 +7,7 @@
 	desc = "Controls a stacking machine... in theory."
 	density = FALSE
 	anchored = TRUE
+	circuit = /obj/item/circuitboard/machine/stacking_unit_console
```
Comment (Fox-McCloud) [inline code/modules/mining/machine_stacking.dm]: You mind fixing the problem with this? below `speed_process` is set to `TRUE`. This machine doesn't need it; it's just a console.

That said, `/obj/machinery/mineral/stacking_machine` hypothetically does (though I'm not sure, anymore, what with `HasProximity`

**Comment by Fox-McCloud** [comment]
Nice, but you're going to cause confusion. The console attempts to locate the stacking machine when it's created; if it can't find one, it immediately deletes itself.

Probably going to be a lot of people going "huh, uh, didn't I build this?"

**Comment by improvedname** [comment]
Just remove the stacking machine old people the future is here O.R.M

**Comment by Fox-McCloud** [comment]
You're confusing the furnace with the stacking machine, I think.

The stacking machine is still used in disposals. It takes in stacks and combined them together inside the stacker, then you can release them with a console. They're still very useful; having one inside cargo could effectively mean that you could have material sheets automatically pulled out of cargo-sorting for you, and you could pretty much leave the recycling windoor open so as soon as the recycling machine released its stacks, it would go back to cargo without fear of it passing through the recycler again.

File: code/modules/research/techweb/all_nodes.dm (hunk 1 above)
Comment (iskyp) [reply code/modules/research/techweb/all_nodes.dm]: i lied

**Comment by iskyp** [comment]
everything is done woohoo
i didn't even need to use the DNM label but i did i guess even though i fixed it in like 2 hours

**Comment by Bawhoppen** [comment]
Finally. One step closer to completely rebuildable station. :+1:x:100:

File: code/modules/mining/machine_stacking.dm [hunk 3]
```diff
@@ -35,6 +33,13 @@
 
 	user << browse(dat, "window=console_stacking_machine")
 
+/obj/machinery/mineral/stacking_unit_console/multitool_act(mob/living/user, obj/item/I)
+	var/obj/item/multitool/M = I
```
Comment (Fox-McCloud) [inline code/modules/mining/machine_stacking.dm]: You usually want to declare vars like this *after* doing a typecheck.

File: code/game/objects/items/circuitboards/machine_circuitboards.dm [hunk 4]
```diff
@@ -864,3 +864,17 @@
 		return
 	transmit = !transmit
 	to_chat(user, "<span class='notice'>You [transmit ? "enable" : "disable"] the board's automatic disposal transmission.</span>")
+
+/obj/item/circuitboard/machine/stacking_unit_console
+	name = "Stacking Machine Console (Machine Board)"
+	build_path = /obj/machinery/mineral/stacking_unit_console
+	req_components = list(
+		/obj/item/stack/sheet/glass = 2,
+		/obj/item/stock_parts/manipulator = 2)
```
Comment (Fox-McCloud) [inline code/game/objects/items/circuitboards/machine_circuitboards.dm]: This is pretty much just a computer; the stacker does all the work. Why not just glass and cable coil?

File: code/modules/mining/machine_stacking.dm (hunk 3 above)
Comment (iskyp) [reply code/modules/mining/machine_stacking.dm]: I tried, but it broke when I did it that way. as in, it stopped working entirely

File: code/game/objects/items/circuitboards/machine_circuitboards.dm (hunk 4 above)
Comment (iskyp) [reply code/game/objects/items/circuitboards/machine_circuitboards.dm]: why didnt I think of that hold on

**Comment by improvedname** [comment]
@fox-mccloud last time i used stacker was at gulag couldnt even drop the minerals i mined cuz it only drops stacks

File: code/modules/mining/machine_stacking.dm [hunk 5]
```diff
@@ -79,6 +85,18 @@
 	if(istype(AM, /obj/item/stack/sheet) && AM.loc == get_step(src, input_dir))
 		process_sheet(AM)
 
+/obj/machinery/mineral/stacking_machine/multitool_act(mob/living/user, obj/item/I)
+	if(istype(I, /obj/item/multitool))
+		var/obj/item/multitool/M = I
+		if(!istype(M.buffer, /obj/machinery/mineral/stacking_unit_console))
+			to_chat(user, "<span class='warning'>The [I] has no linkage data in it's buffer.</span>")
```
Comment (SpaceManiac) [inline code/modules/mining/machine_stacking.dm]: `it's` -> `its`

File: code/modules/mining/machine_stacking.dm (hunk 3 above)
Comment (SpaceManiac) [reply code/modules/mining/machine_stacking.dm]: Don't know why, it's working fine in the other method.

**Review by SpaceManiac — CHANGES_REQUESTED** [review]
I think you may need handling for when `machine` or `CONSOLE` is null, so using unlinked machines doesn't just runtime everywhere.

File: code/modules/mining/machine_stacking.dm (hunk 3 above)
Comment (iskyp) [reply code/modules/mining/machine_stacking.dm]: fucking byond, it works fine now, but last time it was throwing compiler errors at me.
next commit should fix all issues

**Review by SpaceManiac — CHANGES_REQUESTED** [review]

File: code/modules/mining/machine_stacking.dm [hunk 6]
```diff
@@ -19,6 +19,11 @@
 
 /obj/machinery/mineral/stacking_unit_console/ui_interact(mob/user)
 	. = ..()
+
+	if(!machine)
+		to_chat(user, "<span class='notice'>The [src] is not linked to a machine!</span>")
```
[the remark this answers is not in the document]
Comment (SpaceManiac) [reply code/modules/mining/machine_stacking.dm]: `[src]` adds "The" automatically. As is, it will come out "The the stacking machine console".

**Pull request merged**